activitywatch: init at 0.12.2

huantianad opened the pull request:
###### Description of changes
> ActivityWatch is an app that automatically tracks how you spend time on your devices.
>
> It is open source, privacy-first, cross-platform, and a great alternative to services like RescueTime, ManicTime, and WakaTime.
>
> It can be used to keep track of your productivity, time spent on different projects, bad screen habits, or just to understand how you spend your time.

Resolves #97379, https://github.com/ActivityWatch/activitywatch/issues/556.

PR still very WIP and looking for feedback!

Packages 3rd party python modules dependencies under `python3Packages`, as well as `aw-core` and `aw-client` as they're used for making custom watchers. Other parts of the application are packaged normally under `all-packages.nix`

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### TODO
- [ ] Further discuss and iterate on package structure
- [ ] Setup and double check tests for each package.
- [ ] NixOS module which should contain a systemd service for autostart
- [ ] Can we run their integration tests as a NixOS test?

###### Related issues
#59026
[ActivityWatch/activitywatch#556](https://github.com/ActivityWatch/activitywatch/issues/556)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

huantianad committed:
python3Packages.persist-queue: init at 0.8.0

huantianad committed:
python3Packages.timeslot: init at 0.1.2

huantianad committed:
python3Packages.takethetime: init at 0.3.1

huantianad committed:
python3Packages.aw-core: init at 0.5.12

huantianad committed:
python3Packages.aw-client: init at 0.5.11

huantianad committed:
activitywatch: init at 0.12.2

huantianad commented on pkgs/applications/office/activitywatch/default.nix:
```diff
@@ -0,0 +1,157 @@
+
+    postInstall = ''
+      install -D resources/aw-qt.desktop $out/share/applications/aw-qt.desktop
+      install -D resources/aw-qt.desktop $out/etc/xdg/autostart/aw-qt.desktop
```
```suggestion
```
This should be part of a possible module instead of the package itself.

noisersup commented:
What's the state of activitywatch packaging? Is there anything that I could help with?

huantianad commented:
I was planning to wait for 0.12.2 since it should include some small patches to help with packaging, but that seems like it'll take a bit. I'll continue working on this, the main thing to consider are mostly design choices, like making a single package that combines server + qt + watchers, or a nixos module for autostart.

huantianad commented:
I've updated some of the python modules and added some tests, along with general cleanup in this latest change.

I was wondering, is there a good way to combine the individual derivations for each component into one large `activitywatch` derivation for people to install? I think it'd be more userfriendly than having to add `aw-server-rust`, `aw-qt`, etc. all to your packages.

Alternatively, maybe a NixOS module would be a good way to do it to, to allow for autostart?

noisersup commented:
> Alternatively, maybe a NixOS module would be a good way to do it to, to allow for autostart?

Module sound reasonable for that. Not sure if there's something additional that could be done.

huantianad commented:
@noisersup I'll work on the systemd unit, lemme know if you find any bugs! I'm also using this on my daily driver machine, so far it's been pretty good but I don't use it super intensively either, so it's nice to have some more bug testers :)

huantianad committed:
activitywatch: simple symlinkJoin wrapper

huantianad commented:
Hm, so many random hiccups. I decided to create a wrapper for all the different programs that are a part of activitywatch, that simply `symlinkJoin`s them all. The main purpose of this is to both create a single package for users to install, but also a single package for a NixOS module to use in a systemd service.

However, it seems that using `symlinkJoin` doesn't work for `aw-qt`, even though `aw-qt` should search for executables around it starting with `aw-` to load, it seems that it uses the location of the `aw-qt` module, not the executable. 

I'm not sure if it's even worth it to create a single derivation combining all the sub-programs. Should we just direct users to use the NixOS module, or manually add each component to their system packages?

Oro commented:
FYI, in case it's relevant: I cannot seem to build `aw-server-rust` on `nix run nixpkgs#nixpkgs-review pr 202935`, dunno though if that's just me and I'm doing something wrong, if that's the case please feel free to ignore

```
$ git -c fetch.prune=false fetch --no-tags --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0 pull/202935/head:refs/nixpkgs-review/1
remote: Enumerating objects: 4835, done.
remote: Counting objects: 100% (2508/2508), done.
remote: Compressing objects: 100% (14/14), done.
remote: Total 4835 (delta 2494), reused 2504 (delta 2493), pack-reused 2327
Receiving objects: 100% (4835/4835), 3.99 MiB | 7.24 MiB/s, done.
Resolving deltas: 100% (3295/3295), completed with 766 local objects.
From https://github.com/NixOS/nixpkgs
   060e8eee747..93f1a51c1d0  master     -> refs/nixpkgs-review/0
$ git worktree add /home/oro/.cache/nixpkgs-review/pr-202935-2/nixpkgs 93f1a51c1d0ab5697eab9dd4515186bda9da771b
Preparing worktree (detached HEAD 93f1a51c1d0)
Updating files: 100% (34544/34544), done.
HEAD is now at 93f1a51c1d0 Merge pull request #226636 from MatthewCroughan/mc/fix-nixos-shell
$ git merge --no-commit --no-ff f86ddc0e288d54037430daaa599d0c1a01230a82
Auto-merging pkgs/top-level/all-packages.nix
Auto-merging pkgs/top-level/python-packages.nix
Automatic merge went well; stopped before committing as requested
$ nix build --extra-experimental-features nix-command no-url-literals --no-link --keep-going --no-allow-import-from-derivation --option build-use-sandbox relaxed -f /home/oro/.cache/nixpkgs-review/pr-202935-2/build.nix
error: builder for '/nix/store/fr10rwzr6ybwfsr6nqb3sx2bqswxzbbm-aw-server-rust-0.12.2-vendor.tar.gz.drv' failed with exit code 1;
       last 10 log lines:
       > configuring
       > no configure script, doing nothing
       > building
       > source = "git+https://github.com/lawliet89/rocket_cors?rev=54fae07#54fae0701dffbe5df686465780218644ee3fae5f"
       >
       > ERROR: The Cargo.lock contains git dependencies
       >
       > This is currently not supported in the fixed-output derivation fetcher.
       > Use cargoLock.lockFile / importCargoLock instead.
       >
       For full logs, run 'nix log /nix/store/fr10rwzr6ybwfsr6nqb3sx2bqswxzbbm-aw-server-rust-0.12.2-vendor.tar.gz.drv'.
error: 1 dependencies of derivation '/nix/store/ql2a7mkshqyx4igwyi1zafjasy5v15rs-aw-server-rust-0.12.2.drv' failed to build
error: 1 dependencies of derivation '/nix/store/bv9hvsnw82hp1irsm60aqgqgm60i11rv-activitywatch-0.12.2.drv' failed to build
error: 2 dependencies of derivation '/nix/store/cibkb6f0d6m2f2gnxqs3lid8wcxagzm8-review-shell.drv' failed to build

Link to currently reviewing PR:
https://github.com/NixOS/nixpkgs/pull/202935

2 packages failed to build:
activitywatch aw-server-rust

26 packages built:
aw-qt aw-qt.dist aw-watcher-afk aw-watcher-afk.dist aw-watcher-window aw-watcher-window.dist python310Packages.TakeTheTime python310Packages.TakeTheTime.dist python310Packages.aw-client python310Packages.aw-client.dist python310Packages.aw-core python310Packages.aw-core.dist python310Packages.persist-queue python310Packages.persist-queue.dist python310Packages.timeslot python310Packages.timeslot.dist python311Packages.TakeTheTime python311Packages.TakeTheTime.dist python311Packages.aw-client python311Packages.aw-client.dist python311Packages.aw-core python311Packages.aw-core.dist python311Packages.persist-queue python311Packages.persist-queue.dist python311Packages.timeslot python311Packages.timeslot.dist

error: build log of '/nix/store/bv9hvsnw82hp1irsm60aqgqgm60i11rv-activitywatch-0.12.2.drv!*' is not available
error: build log of '/nix/store/a629ypyqp14xaf61pxyd127f56imadm8-activitywatch-0.12.2' is not available
error: build log of '/nix/store/ql2a7mkshqyx4igwyi1zafjasy5v15rs-aw-server-rust-0.12.2.drv!*' is not available
error: build log of '/nix/store/dkgldxyp9inq3v5qyh7sipqqqinyjlbg-aw-server-rust-0.12.2' is not available
$ /nix/store/0rn3w6lr8fn3y502xdzbjkm9vchizkza-nix-2.13.3/bin/nix-shell /home/oro/.cache/nixpkgs-review/pr-202935-2/shell.nix
```

huantianad commented:
I've decided that the `activitywatch` `symlinkJoin` wrapper is worth keeping even if `aw-qt` uses PATH to find it's modules.

I also think that this package is in a state that's ready to be merged. The two main features that I'd like to add would be NixOS module and running integration tests with NixOS test. However, neither are priority and I think it's fine if we add them later.

PS is there anything I should do regarding the `mainProgram` for a `symlinkJoin`? Or any other metadata?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2137

SuperSandro2000 commented on pkgs/top-level/python-packages.nix:
```diff
@@ -11546,6 +11552,8 @@ self: super: with self; {
 
   tailscale = callPackage ../development/python-modules/tailscale { };
 
+  TakeTheTime = callPackage ../development/python-modules/TakeTheTime { };
```
```suggestion
  takethetime = callPackage ../development/python-modules/takethetime { };
```

SuperSandro2000 commented on pkgs/development/python-modules/timeslot/default.nix:
```diff
@@ -0,0 +1,49 @@
+    pytest-cov
+  ];
+
+  checkPhase = ''
+    runHook preCheck
+
+    pytest
+
+    runHook postCheck
+  '';
```
Use pytestCheckHook

SuperSandro2000 commented on pkgs/development/python-modules/aw-core/default.nix:
```diff
@@ -0,0 +1,69 @@
+
+    # Fake home folder for tests that write to $HOME
+    export HOME="$TMP/fake-test-home"
+    pytest tests/
```
use pytestCheckHook

SuperSandro2000 commented on pkgs/development/python-modules/aw-core/default.nix:
```diff
@@ -0,0 +1,69 @@
+    runHook preCheck
+
+    # Fake home folder for tests that write to $HOME
+    export HOME="$TMP/fake-test-home"
```
```suggestion
    export HOME="$TMPDIR"
```

SuperSandro2000 commented on pkgs/development/python-modules/aw-client/default.nix:
```diff
@@ -0,0 +1,62 @@
+
+    # Fake home folder for tests that write to $HOME
+    export HOME="$TMP/fake-test-home"
+    pytest tests/test_requestqueue.py
```
use pytestCheckHook

huantianad replied:
Should still keep the `pythonImportsCheck = [ "aw_client" ];`?

SuperSandro2000 replied:
Yes, please

SuperSandro2000 commented on pkgs/development/python-modules/TakeTheTime/default.nix:
```diff
@@ -0,0 +1,38 @@
+}:
+
+buildPythonPackage rec {
+  pname = "TakeTheTime";
```
```suggestion
  pname = "takethetime";
```
we normalise pname and attrs

SuperSandro2000 commented on pkgs/development/python-modules/TakeTheTime/default.nix:
```diff
@@ -0,0 +1,38 @@
+
+  disabled = pythonOlder "3.6";
+
+  # Tests not included on pypi, use pythonImportsCheck for now.
```
Then fetch from github or so?

SuperSandro2000 commented on pkgs/applications/office/activitywatch/default.nix:
```diff
@@ -0,0 +1,232 @@
+
+    preCheck = ''
+      # Fake home folder for tests that use ~/.cache and ~/.local/share
+      export HOME="$TMP/fake-test-home"
```
```suggestion
      export HOME="$TMPDIR"
```

SuperSandro2000 commented on pkgs/applications/office/activitywatch/default.nix:
```diff
@@ -0,0 +1,232 @@
+
+      # For the actual tray icon, see
+      # https://github.com/ActivityWatch/aw-qt/blob/8ec5db941ede0923bfe26631acf241a4a5353108/aw_qt/trayicon.py#L211-L218
+      install -D media/logo/logo.png $out/lib/python3.10/site-packages/media/logo/logo.png
```
```suggestion
      install -D media/logo/logo.png $out/${python.sitePackages}/media/logo/logo.png
```

SuperSandro2000 commented on pkgs/applications/office/activitywatch/default.nix:
```diff
@@ -0,0 +1,232 @@
+    dontWrapQtApps = true;
+
+    makeWrapperArgs = [
+      "--suffix PATH : \"${lib.makeBinPath [ xdg-utils ]}\""
```
```suggestion
      "--suffix PATH : ${lib.makeBinPath [ xdg-utils ]}"
```

huantianad commented on pkgs/development/python-modules/persist-queue/default.nix:
```diff
@@ -0,0 +1,42 @@
+    # Don't run MySQL tests, as they require a MySQL server running
+    rm persistqueue/tests/test_mysqlqueue.py
+
+    nose2
```
I'm not sure, upstream seems to use `nose2` for testing, are they compatible with each other?

SuperSandro2000 replied:
Sometimes they are but not always. nose2 is old and deprecated and we don't prefer it.

huantianad replied:
I was able to convert `TakeTheTime` to use `pytest`, but it seems persist-queue uses some `nose2` specific tools, i.e. [`nose2.tools.params`](https://github.com/peter-wangxu/persist-queue/blob/00ed89fa7e241df8d3873c356a48939110f0905b/persistqueue/tests/test_queue.py#L11).

huantianad commented on pkgs/applications/office/activitywatch/wrapper.nix:
```diff
@@ -0,0 +1,18 @@
+, extraWatchers ? [ ]
+}:
+
+symlinkJoin {
+  name = "activitywatch-${aw-server-rust.version}";
+  paths = [
+    aw-server-rust.out
+    aw-qt.out
+    aw-watcher-afk.out
+    aw-watcher-window.out
+  ] ++ (lib.forEach extraWatchers (p: p.out));
```
I originally added the symlink join for convenience sake, as the program requires all of these individual parts to function (ie just installing `aw-qt` is useless) and upstream distributes them together. However, I'm not sure if this is standard practice or the correct way to do it, and if not I'm fine with removing it.

SuperSandro2000 replied:
then so be it, sometimes it is easy and just works.

huantianad commented:
@SuperSandro2000 thanks for the review! I've left some comments regarding tests and the symlink join.

jonringer reviewed (changes requested):
Not sure if possible, but it would be nice to add `Cargo.lock` as a patch. It's considered best practice for executables in rust to version control their lock file.

Please follow [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md) and [manual#submitting-changes-making-patches ](https://nixos.org/nixpkgs/manual/#submitting-changes-making-patches) and squash the fix-up commits.

 `git rebase -i` is a powerful command which achieves this, I created a small video demonstrating it's use [here](https://www.youtube.com/watch?v=OR8Q-4PLalc). A more indepth text tutorial can be found [here](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase)

huantianad commented:
Regarding the `Cargo.lock` file, it is indeed already version controlled upstream, and I was previously using it. However, https://github.com/NixOS/nixpkgs/commit/4329de1264ecb7cca8709dc1984b86beb11bcc4e breaks it due to there being a single git dependency, so I switched to putting the `Cargo.lock` file in nixpkgs a la #221716. I haven't read up on all of this yet though, if there's been a better way introduced please let me know and I'll fix it! 

I'll go ahead and rebase, I wasn't sure if I should squash the fixes before they were approved or not, so I erred on the side of doing less. I'll just squash any changes immediately instead of pushing them in the future, thanks for the resources!

SuperSandro2000 reviewed (changes requested):
Please add a pythonImportsCheck to all packages. I think the rest is looking good.

huantianad commented:
@SuperSandro2000 done! Let me know if there's anything I can do to get `pytest` working for `persist-queue`, and if the `symlinkJoin` wrapper is correct (is there anything I need to do in terms of metadata for it?). Thanks for all the help!

ErikBjare commented:
Awesome job, thanks @huantianad! ❤️

not-matthias commented:
Been waiting a while for this. Thank you! :heart:
